### PR TITLE
Add custom-html-template-path config option

### PR DESCRIPTION
**Pull request by wytrych**
This allows for reusing one template file across all tests in use cases
where external scripts or html elements are required for all tests. The
possibility to still use local html files per test file is retained.

Fixes https://github.com/dart-lang/test/issues/39

### CODE REVIEW

**Comment by wytrych** [comment]
@googlebot I signed it!

**Comment by wytrych** [comment]
@jakemac53 Please take a look

**Comment by wytrych** [comment]
I am not sure why the travis build is failing. The `dartanalyzer --fatal-infos --fatal-warnings` command does indeed error for me locally with 8755 errors, but none of this is related to this PR

**Comment by wytrych** [comment]
It looks like something is wrong with my `dartanalyzer` showing all those errors, but nevertheless, I've pulled to dev version of dart and I don't see any hints/errors/warnings in my IntelliJ dart analyzer, so I'm a bit lost on what's wrong.

**Comment by kevmoo** [comment]
@grouma ?

**Comment by kevmoo** [comment]
@wytrych great idea!

File: pkgs/test/lib/src/runner/browser/platform.dart [hunk 1]
```diff
@@ -185,16 +185,23 @@ class BrowserPlatform extends PlatformPlugin
...
-        <head>
-          <title>${htmlEscape.convert(test)} Test</title>
-          $link
-          <script src="packages/test/dart.js"></script>
-        </head>
-        </html>
-      ''', headers: {'Content-Type': 'text/html'});
+      if (_config.customHtmlTemplatePath != null) {
+        var contents = File(_config.customHtmlTemplatePath).readAsStringSync();
+        var processedContents = contents.replaceFirst('{testScript}', link);
```
Comment (jakemac53) [inline pkgs/test/lib/src/runner/browser/platform.dart]: Consider changing this to look for `{{testScript}}` instead of single curlies as I think that is a bit more common for template placeholders.

Comment (jakemac53) [inline pkgs/test/lib/src/runner/browser/platform.dart]: We may also want to allow for a test name placeholder `{{testName}}` (for title tags like below). This would not be required though like the `{{testScript}}` is.

This would also allow the default html file to be itself a template if you wish to do that but it isn't necessary.

File: pkgs/test/lib/src/runner/browser/platform.dart [hunk 2]
```diff
@@ -276,6 +292,15 @@ class BrowserPlatform extends PlatformPlugin
     return suite;
   }
 
+  void checkHtmlCorrectness(String htmlPath, String path) {
```
Comment (jakemac53) [inline pkgs/test/lib/src/runner/browser/platform.dart]: nit: make this private

File: pkgs/test/lib/src/runner/browser/platform.dart (hunk 1 above)
Comment (jakemac53) [inline pkgs/test/lib/src/runner/browser/platform.dart]: Also consider adding a comment here that we already check during the loading phase that this placeholder exists exactly once.

File: pkgs/test/test/runner/browser/runner_test.dart [hunk 3]
```diff
@@ -2,8 +2,9 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-@TestOn('vm')
+import 'dart:convert';
 
+@TestOn('vm')
```
Comment (jakemac53) [inline pkgs/test/test/runner/browser/runner_test.dart]: This annotation needs to move to the first import (or add a library tag and annotate that)

File: pkgs/test/test/runner/browser/runner_test.dart [hunk 4]
```diff
@@ -207,8 +208,145 @@ void main() {
...
+          test.stdout,
+          containsInOrder([
+            '-1: compiling test.dart [E]',
+            'Failed to load "test.dart": "test.html" must contain '
+                '<script src="packages/test/dart.js"></script>.'
+          ]));
+      await test.shouldExit(1);
+    }, tags: 'chrome');
+
     // TODO(nweiz): test what happens when a test file is unreadable once issue
```
Comment (jakemac53) [inline pkgs/test/test/runner/browser/runner_test.dart]: nit: move this back to the bottom of the group or delete it

File: pkgs/test_core/lib/src/runner/configuration.dart [hunk 5]
```diff
@@ -43,6 +43,10 @@ class Configuration {
   bool get help => _help ?? false;
   final bool _help;
 
+  /// Custom HTML template file
+  String get customHtmlTemplatePath => _customHtmlTemplatePath;
```
Comment (jakemac53) [inline pkgs/test_core/lib/src/runner/configuration.dart]: Just make the final variable public here - it already has no public setter.

The other ones have getters because they do some additional logic :)

File: pkgs/test/README.md [hunk 6]
```diff
@@ -518,6 +521,37 @@ the following HTML file:
...
 ```
 
+### Providing a custom HTML template
+
+If you want to share the same HTML file across all tests, you can provide a
+`custom-html-template-path` configuration option to your configuration file.
+This file should follow the rules above, except that instead of the link tag
+add exactly one `{testScript}` in the place where you want the template processor to insert it.
+
+The template can't end with `.html` since the test loader will not able to process it and an error will be thrown.
```
Comment (jakemac53) [inline pkgs/test/README.md]: I am a bit confused by why it errors - does it glob for all html files at some point? I wouldn't expect the test loader to ever try to load this file.

File: pkgs/test/lib/src/runner/browser/platform.dart (hunk 1 above)
Comment (wytrych) [reply pkgs/test/lib/src/runner/browser/platform.dart]: I'm not sure I understand the `This would also allow the default html file to be itself a template ` bit. You mean the HTML that is in the else branch, so it would be processed in the same way as the user provided, thus simplifying the logic?

Comment (jakemac53) [reply pkgs/test/lib/src/runner/browser/platform.dart]: Yes exactly, it would simplify/unify the logic a bit. It would also be (negligibly) slower since it has to reach out to disk. You would likely use `Isolate.resolvePackageUri` to get the absolute file uri for the template file.

But again this is just a suggestion and not a requirement, I am fine with it as is :).

File: pkgs/test/README.md (hunk 6 above)
Comment (wytrych) [reply pkgs/test/README.md]: Sorry, this might be my bad writing. It will throw because I made it so in https://github.com/dart-lang/test/pull/1127/files/343af3e063c748e1323177e82854dbb0aaf1a33f#diff-b9a58c9d38842cb742b063d60ff6fb18R257 But the point of this is that if it is an `.html` file it will be served directly by shelf and will not be handled by the wrapper in `platform.dart`

Comment (jakemac53) [reply pkgs/test/README.md]: This file should only ever need to be read from disk as a part of the `platform.dart` code though (it doesn't go through the shelf handler) - and it should be fine for shelf to serve it directly even if it was requested for some reason?

**Comment by jakemac53** [comment]
Oh also one quick note please don't force-push your branch just push commits - we will do a squash and merge.

This makes it easier to review changes :).

File: pkgs/test/README.md (hunk 6 above)
Comment (wytrych) [reply pkgs/test/README.md]: Hmm, so the issue I was having is that the `suiteUrl` is generated in the `load` method by taking the dart file name and replacing `.dart` with `.html`. If shelf then tries to load that html file but it doesn't exist on disk, it will go into the wrapper handler and generate the file.
However if it does, it will be served directly, so it is not possible to process it, so it can't really be used as a template anymore. I came to the conclusion that it's less confusing if the template can't be an html file, because then it would be hard to understand without looking at the internal mechanics why it isn't processed. I decided then to disallow that during the config parsing phase.

Sure, it is kind of an edge case since the template file would have to be named like the test file, but I think it's nice to separate the old mechanics of having a custom html file for the test and having a template file. But if you think it should be allowed, then I'll remove that check.

Comment (jakemac53) [reply pkgs/test/README.md]: Ah ok I get where the error came from now - IMO I don't think that is a common enough case to block all `.html` file extensions though, cc @natebosch wdyt?

Comment (natebosch) [reply pkgs/test/README.md]: +1. If anything we should check that it doesn't match exactly the `.html` file for an existing test case.

Comment (jakemac53) [reply pkgs/test/README.md]: Ah ya we could change the check to that instead - I like that.

**Comment by wytrych** [comment]
Thanks @jakemac53 and @natebosch  I pushed fixes

File: pkgs/test/lib/src/runner/browser/static/default.html.tpl [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>{{testName}}</title>
```
Comment (jakemac53) [inline pkgs/test/lib/src/runner/browser/static/default.html.tpl]:
```suggestion
    <title>{{testName}} Test</title>
```

To match exactly what it was before

**Comment by jakemac53** [comment]
Going to let @grouma take a pass as well before merging (I think he is back Monday)

File: pkgs/test/lib/src/runner/browser/platform.dart [hunk 8]
```diff
@@ -126,7 +128,11 @@ class BrowserPlatform extends PlatformPlugin
   /// Mappers for Dartifying stack traces, indexed by test path.
   final _mappers = <String, StackTraceMapper>{};
 
+  /// The default template for html tests
```
Comment (grouma) [inline pkgs/test/lib/src/runner/browser/platform.dart]: [nit] End in a period. (here and elsewhere)


https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

**Comment by wytrych** [comment]
Thanks @grouma, I pushed the fixes.

**Comment by jakemac53** [comment]
@grouma are there other changes we want to roll into this release?

**Comment by wytrych** [comment]
Do you have an idea on when you will be releasing?

**Comment by jakemac53** [comment]
It should be soon but we have a couple other prs we would like to merge, https://github.com/dart-lang/test/pull/1129 and ~https://github.com/dart-lang/test/pull/1131~

**Comment by wytrych** [comment]
Cool, thanks!

**Comment by grouma** [comment]
Just published.
